### PR TITLE
JCLOUDS-1005: Backblaze B2 object operations

### PR DESCRIPTION
**Pull request by andrewgaul**

### CODE REVIEW

**Comment by andrewgaul** [comment]
Updated with all object operations.

File: b2/src/main/java/org/jclouds/b2/binders/UploadFileBinder.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+import com.google.common.net.PercentEscaper;
+
+public final class UploadFileBinder implements MapBinder {
+   public static final PercentEscaper escaper = new PercentEscaper("._-/~!$'()*;=:@", false);
+
+   @Override
+   public <R extends HttpRequest> R bindToRequest(R request, Map<String, Object> postParams) {
+      UploadUrlResponse uploadUrl = (UploadUrlResponse) postParams.get("uploadUrl");
+      B2Object b2Object = (B2Object) postParams.get("b2Object");
+      Preconditions.checkArgument(b2Object.fileId() == null, "fileId must be null, was %s", b2Object.fileId());
```
Comment (zack-shoylev) [inline b2/src/main/java/org/jclouds/b2/binders/UploadFileBinder.java]: One way we handle this (in Openstack providers) if to have 2 separate classes, one that is used to create an object, and one that is used when reading an object. Thus fields that can only be used for updating or reading cannot be specified incorrectly. AutoValue makes it easier to implement - you implement a regular autovalue for the "read" version of the class, and you implement an AutoValue with an AutoValue builder for the create/update (update could  be a separate class as well).

I also think that doing client-side validation checks is largely empty work and removing them helps to (sometimes completely) eliminate binders and mappers.

Thoughts?

File: b2/src/main/java/org/jclouds/b2/domain/B2Object.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+   public static B2Object createResponse(@Nullable String fileId, String fileName, String contentSha1, Map<String, String> fileInfo, @Nullable Payload payload, @Nullable Date uploadTimestamp) {
+      Preconditions.checkArgument(contentSha1.length() == Hashing.sha1().bits() / 4,
+            "Expected hash length of %s, was: %s", Hashing.sha1().bits() / 4, contentSha1.length());
+      return new AutoValue_B2Object(fileId, fileName, contentSha1, fileInfo, payload, uploadTimestamp, null, null, null, null, null);
+   }
+
+   @SerializedNames({"fileId", "fileName", "accountId", "bucketId", "contentLength", "contentSha1", "contentType", "fileInfo", "action", "uploadTimestamp", "payload"})
+   public static B2Object createHeadResponse(String fileId, String fileName, String accountId, String bucketId, long contentLength, String contentSha1, String contentType, Map<String, String> fileInfo, Action action, long uploadTimestamp, @Nullable Payload payload) {
+      return new AutoValue_B2Object(fileId, fileName, contentSha1, fileInfo, null, new Date(uploadTimestamp), action, accountId, bucketId, contentLength, contentType);
+   }
```
Comment (nacx) [inline b2/src/main/java/org/jclouds/b2/domain/B2Object.java]: In all factory methods, enforce immutable collections. Use ImmutableMap.copyOf for fileInfo.

File: b2/src/main/java/org/jclouds/b2/domain/UploadFileResponse.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+   public abstract String accountId();
+   public abstract String bucketId();
+   public abstract long contentLength();
+   public abstract String contentSha1();
+   public abstract String contentType();
+   public abstract Map<String, String> fileInfo();
+
+   @SerializedNames({"fileId", "fileName", "accountId", "bucketId", "contentLength", "contentSha1", "contentType", "fileInfo"})
+   public static UploadFileResponse create(String fileId, String fileName, String accountId, String bucketId, long contentLength, String contentSha1, String contentType, Map<String, String> fileInfo) {
+      return new AutoValue_UploadFileResponse(fileId, fileName, accountId, bucketId, contentLength, contentSha1, contentType, fileInfo);
```
Comment (nacx) [inline b2/src/main/java/org/jclouds/b2/domain/UploadFileResponse.java]: Immutable fileinfo

File: b2/src/test/java/org/jclouds/b2/features/ObjectApiLiveTest.java [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+      } finally {
+         if (uploadFile != null) {
+            objectApi.deleteFileVersion(uploadFile.fileName(), uploadFile.fileId());
+         }
+         bucketApi.deleteBucket(response.bucketId());
+      }
+   }
+
+   @Test(groups = "live")
+   public void testDownloadFileByNameNonExistent() throws Exception {
```
Comment (nacx) [inline b2/src/test/java/org/jclouds/b2/features/ObjectApiLiveTest.java]: In general, I don't think we need to "live" test stuff like this. We are creating unnecessary resources just to validate a jclouds behavior that does not give us any value of "how aligned with the provider api we are". The purpose of this test is to verify that we propagate and exception, and that is alreaydy covered in the mock tests.

File: b2/src/main/java/org/jclouds/b2/features/ObjectApi.java [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+   B2ObjectList listFileNames(@PayloadParam("bucketId") String bucketId, @PayloadParam("startFileName") @Nullable String startFileName, @PayloadParam("maxFileCount") @Nullable Integer maxFileCount);
+
+   @Named("b2_list_file_versions")
+   @GET
+   @Path("/b2api/v1/b2_list_file_versions")
+   @MapBinder(BindToJsonPayload.class)
+   @RequestFilters(RequestAuthorization.class)
+   @Consumes(APPLICATION_JSON)
+   @Produces(APPLICATION_JSON)
+   B2ObjectList listFileVersions(@PayloadParam("bucketId") String bucketId, @PayloadParam("startFileId") @Nullable String startFileId, @PayloadParam("startFileName") @Nullable String startFileName, @PayloadParam("maxFileCount") @Nullable Integer maxFileCount);
```
Comment (nacx) [inline b2/src/main/java/org/jclouds/b2/features/ObjectApi.java]: Does it make sense to add a null fallback on 404 for these list methods?

File: b2/src/main/java/org/jclouds/b2/features/ObjectApi.java [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+   UploadFileResponse uploadFile(@PayloadParam("uploadUrl") UploadUrlResponse uploadUrl, @PayloadParam("b2Object") B2Object b2Object);
+
+   @Named("b2_delete_file_version")
+   @POST
+   @Path("/b2api/v1/b2_delete_file_version")
+   @MapBinder(BindToJsonPayload.class)
+   @RequestFilters(RequestAuthorization.class)
+   @Consumes(APPLICATION_JSON)
+   @Produces(APPLICATION_JSON)
+   DeleteFileResponse deleteFileVersion(@PayloadParam("fileName") String fileName, @PayloadParam("fileId") String fileId);
```
Comment (nacx) [inline b2/src/main/java/org/jclouds/b2/features/ObjectApi.java]: Does it make sense to add a null fallback on a 404?

File: b2/src/main/java/org/jclouds/b2/features/ObjectApi.java [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+
+   @Named("b2_get_file_info")
+   @POST
+   @Path("/b2api/v1/b2_get_file_info")
+   @MapBinder(BindToJsonPayload.class)
+   @RequestFilters(RequestAuthorization.class)
+   @Consumes(APPLICATION_JSON)
+   @Produces(APPLICATION_JSON)
+   B2Object getFileInfo(@PayloadParam("fileId") String fileId);
```
Comment (nacx) [inline b2/src/main/java/org/jclouds/b2/features/ObjectApi.java]: Does it make sense to add a null fallback on a 404?

File: b2/src/main/java/org/jclouds/b2/features/ObjectApi.java [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+@BlobScope(CONTAINER)
+public interface ObjectApi {
+   @Named("b2_get_upload_url")
+   @POST
+   @Path("/b2api/v1/b2_get_upload_url")
+   @RequestFilters(RequestAuthorization.class)
+   @MapBinder(BindToJsonPayload.class)
+   @Consumes(APPLICATION_JSON)
+   @Produces(APPLICATION_JSON)
+   UploadUrlResponse getUploadUrl(@PayloadParam("bucketId") String bucketId);
```
Comment (nacx) [inline b2/src/main/java/org/jclouds/b2/features/ObjectApi.java]: Does it make sense to add a null fallback on a 404?

File: b2/src/test/java/org/jclouds/b2/features/ObjectApiMockTest.java [hunk 9]
```diff
@@ -0,0 +1,481 @@
...
+         assertThat(request.getMethod()).isEqualTo("GET");
+         assertThat(request.getPath()).isEqualTo("/b2api/v1/b2_authorize_account");
+
+         request = server.takeRequest();
+         assertThat(request.getMethod()).isEqualTo("GET");
+         assertThat(request.getPath()).isEqualTo("/file/BUCKET_NAME/typing_test.txt");
+      } finally {
+         server.shutdown();
+      }
+   }
```
Comment (nacx) [inline b2/src/test/java/org/jclouds/b2/features/ObjectApiMockTest.java]: Add mock tests for the download methods that declare fallbacks.

**Comment by nacx** [comment]
This looks great @andrewgaul! Just a few minor comments :)

File: b2/src/main/java/org/jclouds/b2/domain/B2Object.java (hunk 2 above)
Comment (andrewgaul) [inline b2/src/main/java/org/jclouds/b2/domain/B2Object.java]: Done.

File: b2/src/main/java/org/jclouds/b2/domain/UploadFileResponse.java (hunk 3 above)
Comment (andrewgaul) [inline b2/src/main/java/org/jclouds/b2/domain/UploadFileResponse.java]: Done.

File: b2/src/test/java/org/jclouds/b2/features/ObjectApiLiveTest.java (hunk 4 above)
Comment (andrewgaul) [inline b2/src/test/java/org/jclouds/b2/features/ObjectApiLiveTest.java]: Removed.

File: b2/src/main/java/org/jclouds/b2/features/ObjectApi.java (hunk 6 above)
Comment (andrewgaul) [inline b2/src/main/java/org/jclouds/b2/features/ObjectApi.java]: I changed `ParseB2ErrorFromJsonContent` to throw `KeyNotFoundException` -- what do you think?

File: b2/src/main/java/org/jclouds/b2/features/ObjectApi.java (hunk 5 above)
Comment (andrewgaul) [inline b2/src/main/java/org/jclouds/b2/features/ObjectApi.java]: Oddly, B2 returns 200 when issuing `listFileVersions` against an already-deleted bucket.

File: b2/src/main/java/org/jclouds/b2/features/ObjectApi.java (hunk 7 above)
Comment (andrewgaul) [inline b2/src/main/java/org/jclouds/b2/features/ObjectApi.java]: Done.

File: b2/src/test/java/org/jclouds/b2/features/ObjectApiMockTest.java (hunk 9 above)
Comment (andrewgaul) [inline b2/src/test/java/org/jclouds/b2/features/ObjectApiMockTest.java]: Done.

File: b2/src/main/java/org/jclouds/b2/features/ObjectApi.java (hunk 8 above)
Comment (andrewgaul) [inline b2/src/main/java/org/jclouds/b2/features/ObjectApi.java]: I'm not sure about this -- users will likely directly feed the result of `getUploadUrl` into `uploadFile`.  Throwing `ContainerNotFoundException` in the former gives more information that the `NullPointerException` from the latter.

File: b2/src/main/java/org/jclouds/b2/binders/UploadFileBinder.java (hunk 1 above)
Comment (andrewgaul) [inline b2/src/main/java/org/jclouds/b2/binders/UploadFileBinder.java]: Done.

**Comment by andrewgaul** [comment]
Addressed all outstanding review comments.

**Comment by zack-shoylev** [comment]
Looks good to me!

**Comment by nacx** [comment]
+1!

**Pull request closed without merging**